### PR TITLE
refactor(agui): redesign event dispatching with strategy pattern and deferred queue

### LINKED ISSUE #1347: [Feature]: refactor ag-ui adapter
**Description:**

**1. Background & Current Pain Points**
The current `AguiAgentAdapter` is a core component bridging AgentScope and the AG-UI protocol. However, as the framework evolves, the current monolithic implementation exposes several architectural bottlenecks:
- **Poor Extensibility (If-Else Mountain):** The `convertEvent` method relies on deeply nested `if-else` blocks to parse different `ContentBlock` types. Adding support for any new block type requires modifying this core method, violating the Open-Closed Principle (OCP).
- **State Management Complexity:** It uses `EventConversionState` which maintains multiple `Set<String>` collections to track the lifecycle (`hasStarted`, `hasEnded`) of messages and tools. This makes handling interruption scenarios (e.g., Tool call interrupting Text) extremely complicated.
- **Hardcoded Conversion Logic:** The mapping from `ContentBlock` to AG-UI events is rigidly embedded within the adapter. Developers cannot inject custom parsing logic or override default block processing strategies without modifying the core framework code.

**2. Proposed Architectural Changes**
We propose refactoring the adapter using the **Strategy Pattern** combined with a **Deferred Queue** for strict lifecycle closure.

Here is the topology comparison before and after the refactoring:

```text
🔴 Before: Monolithic Architecture (Nested If-Else)
[ Agent Event Stream ] 
       │
       ▼
┏━━━━━━━━━━━━━━━━━━━━━ AguiAgentAdapter ━━━━━━━━━━━━━━━━━━━━━┓
┃ 1. Global State: EventConversionState (Growing Set collections)
┃ 2. Monolithic Parsing: convertEvent(Event)                 ┃
┃    ├── if (EventType == REASONING/SUMMARY)                 ┃
┃    │    ├── for (ContentBlock)                             ┃
┃    │    │    ├── if (instanceof TextBlock)      => logic   ┃
┃    │    │    ├── else if (instanceof Thinking)  => logic   ┃
┃    │    │    └── else if (instanceof ToolUse)   => logic   ┃
┃    └── else if (EventType == TOOL_RESULT) ...              ┃
┃ 3. Manual Cleanup: Loop through all Sets to fix missing Ends
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛

🟢 After: Strategy Router + Deferred Queue
[ Agent Event Stream ]
       │
       ▼
┏━━━━━━━━━━━━━━━━━━━━━ AguiAgentAdapter ━━━━━━━━━━━━━━━━━━━━━┓
┃ 1. Dispatcher: converters.get(block.getClass()).convert()  ┃
┃         ├─▶ TextBlockConverter       ────────┐             ┃
┃         ├─▶ ThinkingBlockConverter   ────────┼─▶ Emit Event┃
┃         ├─▶ ToolUseBlockConverter    ────────┤   or        ┃
┃         ├─▶ ToolResultBlockConverter ────────┤   Defer End ┃
┃         └─▶ [User Custom Converters] ────────┘             ┃
┃ 2. Context: StreamContext (O(1) Active tracking & Deferred Queue)
┃ 3. Auto-Closure: Flux.fromIterable(ctx.flushAllRemainingDeferred())
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
```

**3. Expected Benefits**
- **Zero-Modification Extensibility:** Future support for Image/Audio/Video blocks will only require adding a new `Converter` class.
- **Strict AG-UI Compliance:** The Deferred Queue guarantees that every `Start` event strictly pairs with an `End` event, even during exceptions or interruptions.
- **Pluggable Custom Converters:** By exposing `customConverters` in the configuration, developers can seamlessly inject their own strategies to override default behaviors or support domain-specific block processing, transforming the adapter into a highly customizable microkernel architecture.

### PR DESCRIPTION
**Pull request by jujn**
**Description:**
Closes #1347, #1382 

This PR implements the architectural refactoring proposed in #1347, bringing a strategy-based dispatcher to the AG-UI adapter and enabling pluggable custom converters. Additionally, it fixes the lifecycle mismatch bug mentioned in #1382.

#### Code Changes & Architecture
1. **Refactored `AguiAgentAdapter`:** Replaced the massive `if-else` block with a `BlockEventConverter` strategy interface. The adapter now acts as a clean router (dispatching based on Block class).
2. **Introduced `StreamContext`:** Replaced the heavy `Set`-based state tracking with a lightweight context. It uses a **Deferred Queue** mechanism (delaying `End` events until termination or interruption), resolving lifecycle mismatch bugs.
3. **Added `BlockEventConverter` Implementations:** Logic for `TextBlock`, `ThinkingBlock`, `ToolUseBlock`, and `ToolResultBlock` are now beautifully isolated in their own classes.
4. **Introduced Pluggable `customConverters`:** To support future broader customization needs without modifying the core framework, we introduced new configuration mechanisms:
    - **Core Config (`AguiAdapterConfig`)**: Added `customConverters` to allow developers to register their own `BlockEventConverter` implementations.
    - **Spring Boot Auto-Configuration (`agui-spring-boot-starter`)**: Modified `AgentscopeAguiMvcAutoConfiguration` and `AgentscopeAguiWebFluxAutoConfiguration` by utilizing `ObjectProvider<BlockEventConverter<?>>`. This allows the framework to automatically collect any custom converter beans defined in the Spring context (e.g., using `@Component`), transforming the adapter into a highly extensible microkernel architecture.
5. **Updated Tests:** For AguiAgentAdapterTest.java, all existing unit tests pass smoothly(just modified one), proving backward compatibility. And added new tests to cover the remaining newly added code.

> **[📝 Multi-Model Log Comparison]**
> 
> To ensure that the refactored AguiAgentAdapter behaves consistently and correctly across different LLMs, I evaluated the AG-UI event outputs before and after the refactoring by running 12 test cases(See the attachment agui_test.js) across each of the 5 different models (qwen3.5-plus, gpt-5.2, kimi-k2.6, MiniMax-M2.5, deepseek-v4-pro, glm-5).
> 
> *Testing context:* The tests were conducted using `agentscope-examples/integration/agui` with reasoning events enabled. (Test scripts and full logs are attached).
> *Result:* After a rigorous dual review by human and AI, **no issues have been found so far post-refactoring.**
> 
> **Appendix: AG-UI Log Review Criteria**
> The logs were meticulously audited against the following standards:
> 1. **Lifecycle Completeness:** Ensure the complete lifecycle (`start`, `content`, `end`) for reasoning events, tool call events, and text events (tool calls also include a `result` event). The `messageId` or `toolCallId` must remain exactly consistent throughout each lifecycle.
> 2. **Strict Event Ordering:** Ensure events follow a rigorous sequence (e.g., a reasoning event must fully end before a text event starts; a reasoning event must fully end before a tool event starts).
> 3. **No Duplicate Events:** Check for redundant `start` or `end` events within each test case (e.g., no multiple reasoning start/end events sharing the identical `messageId`).
> 4. **Content Integrity:** Verify that concatenated `content` events form a complete message without any truncation or interruption.
> 5. **Focus:** Note that the "before" logs may contain pre-existing framework bugs. The primary focus of this audit is verifying the absolute correctness of the "after" logs (post-refactoring).

[测试脚本及日志.zip](https://github.com/user-attachments/files/28177133/default.zip)

### CODE REVIEW

File: agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/AguiAgentAdapter.java [hunk 1]
```diff
@@ -66,17 +70,25 @@ public class AguiAgentAdapter {
...
+        this.config = Objects.requireNonNull(config);
         this.messageConverter = new AguiMessageConverter();
+
+        // Register block conversion strategies
+        converters.put(TextBlock.class, new TextBlockConverter());
+        converters.put(ThinkingBlock.class, new ThinkingBlockConverter());
+        converters.put(ToolUseBlock.class, new ToolUseBlockConverter());
+        converters.put(ToolResultBlock.class, new ToolResultBlockConverter());
+        converters.put(CustomBlock.class, new CustomBlockConverter());
     }
```
[the remark this answers is not in the document]
Comment (jujn) [reply agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/AguiAgentAdapter.java]: After the review is approved, I will proceed to split and refactor AguiAgentAdapterTest.

**Comment by jujn** [comment]
@LearningGp 
Could you please prioritize reviewing this PR when you have time? This is very important for the future maintainability and extensibility of the ag-ui module. Thanks very mach.

**Comment by LearningGp** [comment]
For PRs related to ag-ui, please cc @chickenlj  for a high-level review to avoid potential conflicts and misalignment.

**Review by AgentScopeJavaBot — COMMENTED** [review]
🤖 AI Review

This PR delivers a well-architected refactoring of the AG-UI adapter, replacing a monolithic if-else chain with a clean strategy pattern (`BlockEventConverter`) and introducing a `StreamContext` with deferred-end-event queue to resolve lifecycle ordering issues (#1382). The per-subscription state isolation via `StreamContext` is a significant improvement over the old instance-field approach, making the adapter safe for concurrent/retry subscriptions. Test coverage is strong with 30+ test methods including state-leak and custom-converter-override tests. The Spring Boot integration correctly wires custom converter beans via `ObjectProvider`. A few defensive-programming gaps and missing direct tests for `StreamContext` prevent a clean approval.

File: agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/AguiAgentAdapter.java [hunk 2]
```diff
@@ -102,386 +117,74 @@ public Flux<AguiEvent> run(RunAgentInput input) {
...
-                                    toolCallId,
-                                    result,
-                                    "tool",
-                                    msg.getId()));
-                    state.endToolCall(toolCallId);
-                }
+    @SuppressWarnings("unchecked")
+    private List<AguiEvent> processEvent(Event event, StreamContext ctx) {
+        // Dispatch each content block to its corresponding converter
+        for (ContentBlock block : event.getMessage().getContent()) {
```
Comment (AgentScopeJavaBot) [inline agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/AguiAgentAdapter.java:166]: [major] **Potential NPE on upstream null message.** `event.getMessage()` can be null — the `Event` constructor does not enforce `Objects.requireNonNull` on the `message` parameter. If a malformed agent stream emits an event with a null message, this line will throw `NullPointerException`, crashing the reactive pipeline without a meaningful error message.

Suggested guard:
```java
if (event.getMessage() == null || event.getMessage().getContent() == null) {
    return ctx.getAndClearEmittedEvents();
}
```

File: agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/AguiAgentAdapter.java [hunk 3]
```diff
@@ -66,17 +62,36 @@ public class AguiAgentAdapter {
     private final Agent agent;
     private final AguiAdapterConfig config;
     private final AguiMessageConverter messageConverter;
+    private final Map<Class<?>, BlockEventConverter<?>> converters = new HashMap<>();
```
Comment (AgentScopeJavaBot) [inline agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/AguiAgentAdapter.java]: [minor] **`converters` map should be made unmodifiable after construction.** This `HashMap` is populated only during the constructor (via `registerDefault` + `putAll`) and subsequently used in read-only fashion by `processEvent`. Wrapping it in `Collections.unmodifiableMap()` after initialization would make the immutability intent explicit, prevent accidental mutation in future refactors, and document the thread-safety contract for concurrent `run()` calls.

File: agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/AguiAgentAdapter.java [hunk 4]
```diff
@@ -102,386 +117,74 @@ public Flux<AguiEvent> run(RunAgentInput input) {
...
-                                    "tool",
-                                    msg.getId()));
-                    state.endToolCall(toolCallId);
-                }
+    @SuppressWarnings("unchecked")
+    private List<AguiEvent> processEvent(Event event, StreamContext ctx) {
+        // Dispatch each content block to its corresponding converter
+        for (ContentBlock block : event.getMessage().getContent()) {
+            BlockEventConverter<ContentBlock> converter =
+                    (BlockEventConverter<ContentBlock>) converters.get(block.getClass());
```
Comment (AgentScopeJavaBot) [inline agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/AguiAgentAdapter.java:168]: [minor] **Exact-class dispatch may surprise custom converter authors.** `converters.get(block.getClass())` performs exact class matching — a converter registered for a parent `ContentBlock` subtype will **not** match subclass instances. Since built-in types (`TextBlock`, `ToolUseBlock`, etc.) are `final`, this is fine internally. However, users who register custom converters for non-final block types may expect `instanceof`-style dispatch. Consider documenting this constraint in the `BlockEventConverter` Javadoc, or switching to a linear scan with `isAssignableFrom` if subclass support is desired.

File: agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/StreamContext.java [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Context holder for the AG-UI stream pipeline.
+ * Manages the emission of immediate events and the queuing of deferred end events
+ * to ensure strict adherence to the AG-UI lifecycle protocol.
```
Comment (AgentScopeJavaBot) [inline agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/StreamContext.java:30]: [minor] **Missing dedicated unit tests for `StreamContext`.** This class encapsulates non-trivial state-machine logic: deferred event queuing, active/finished tracking across three dimensions (text, reasoning, tool), and flush semantics on completion/error paths. While `AguiAgentAdapterTest` exercises these indirectly, direct unit tests for `StreamContext` would provide faster feedback and catch edge cases (e.g., flushing a non-existent deferred event, double-remove of an active ID, interleaved flush calls) more precisely.

File: agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/strategy/ToolResultBlockConverter.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    public void convert(ToolResultBlock block, Event event, StreamContext ctx) {
+        String toolCallId =
+                block.getId() != null && !block.getId().isBlank()
+                        ? block.getId()
+                        : UUID.randomUUID().toString();
+
+        String result = extractToolResultText(block);
+
+        // Closing Start/End Phase
+        if (ctx.isToolActive(toolCallId)) {
```
Comment (AgentScopeJavaBot) [inline agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/strategy/ToolResultBlockConverter.java:52]: [minor] **Event ordering change vs. old code — verify frontend compatibility.** In the old adapter, the event sequence was `ToolCallResult → ToolCallEnd` (flush). The new code emits `ToolCallEnd` (via `flushEndEvent` at line 53) **before** `ToolCallResult` (line 68). This matches the AG-UI protocol enum ordering (`TOOL_CALL_END` before `TOOL_CALL_RESULT`) and appears intentional per the lifecycle-fix goal of #1382. However, if any existing frontend consumer was depending on the old ordering, this is a breaking behavioral change. Consider calling this out explicitly in the PR description or release notes.

File: agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/strategy/ToolResultBlockConverter.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+            String toolName =
+                    block.getName() != null && !block.getName().isBlank()
+                            ? block.getName()
+                            : "unknown";
+            ctx.emit(
+                    new AguiEvent.ToolCallStart(
+                            ctx.getThreadId(), ctx.getRunId(), toolCallId, toolName));
+            ctx.emit(new AguiEvent.ToolCallEnd(ctx.getThreadId(), ctx.getRunId(), toolCallId));
+        }
+
```
Comment (AgentScopeJavaBot) [inline agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/strategy/ToolResultBlockConverter.java:66]: [nit] **Fallback backfill emits `ToolCallEnd` before `ToolCallResult` — same ordering as main path.** In the fallback branch (lines 55–66), the sequence is `ToolCallStart → ToolCallEnd → ToolCallResult`. This is consistent with the main path and the AG-UI protocol enum order. Just confirming this is intentional — no action needed.

File: agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/AguiAdapterConfig.java [hunk 8]
```diff
@@ -200,6 +248,22 @@ public Builder defaultAgentId(String defaultAgentId) {
...
+         * Register a custom block event converter.
+         *
+         * <p>This allows users to override the default conversion strategies for specific
+         * ContentBlock types, enabling deep customization of AG-UI event generation.
+         *
+         * @param converter The custom converter strategy
+         * @return This builder
+         */
+        public Builder registerConverter(BlockEventConverter<?> converter) {
+            if (converter != null && converter.supportedBlockType() != null) {
```
Comment (AgentScopeJavaBot) [inline agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/AguiAdapterConfig.java:261]: [nit] **Silent overwrite in `registerConverter`.** If two converters are registered for the same `supportedBlockType()`, the second one silently overwrites the first without any log or warning. This is standard `Map.put` behavior and acceptable, but a debug-level log message when an override occurs could help users troubleshoot misconfiguration.

File: agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/AguiAgentAdapter.java [hunk 9]
```diff
@@ -102,386 +117,74 @@ public Flux<AguiEvent> run(RunAgentInput input) {
                             StreamOptions.builder()
                                     .eventTypes(EventType.ALL)
                                     .incremental(true)
+                                    .includeActingChunk(config.isEnableActingChunk())
                                     .build();
```
Comment (AgentScopeJavaBot) [inline agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/AguiAgentAdapter.java:127]: [praise] **Clean per-subscription state isolation.** Creating a fresh `StreamContext` inside `Flux.defer()` ensures each subscription gets its own state — fixing the old instance-field-based state leak that could corrupt concurrent or retry subscriptions. Well-designed fix for the underlying issue.

File: agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/StreamContext.java [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+
+    /**
+     * Registers an end event to be emitted later.
+     * This decouples the start logic from the termination logic.
+     *
+     * @param id       The prefixed identifier for the component
+     * @param endEvent The end event to defer
+     */
+    public void deferEndEvent(String id, AguiEvent endEvent) {
+        deferredEndEvents.put(id, endEvent);
```
Comment (AgentScopeJavaBot) [inline agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/StreamContext.java:115]: [praise] **Elegant deferred-end-event queue.** The `LinkedHashMap<String, AguiEvent>` keyed by `prefix + id` is a clean mechanism for ensuring `End` events are always emitted in the correct lifecycle position — either when the logical unit finishes naturally, when interrupted by a new unit, or at stream termination/error. The prefix scheme (`text:`, `reasoning:`, `tool:`) correctly prevents key collisions between different block types sharing the same message ID.

File: agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/strategy/BlockEventConverter.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ * Strategy interface for converting different types of ContentBlock into AG-UI events.
+ *
+ * @param <T> The specific type of ContentBlock this converter handles.
+ */
+public interface BlockEventConverter<T extends ContentBlock> {
+
+    /**
+     * Retrieves the specific type of ContentBlock that this converter can process.
+     *
```
Comment (AgentScopeJavaBot) [inline agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/strategy/BlockEventConverter.java:31]: [praise] **Well-designed strategy interface.** Three methods (`supportedBlockType`, `isApplicable`, `convert`) provide a clean separation: type routing, event filtering, and conversion logic. This is a textbook application of the Strategy pattern that eliminates the old if-else chain and makes each block type's handling independently testable.

File: agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/AguiAgentAdapter.java (hunk 2 above)
Comment (AgentScopeJavaBot) [reply agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/AguiAgentAdapter.java:166]: ✅ Verified as addressed in `6f36b22`: processEvent now has explicit null guard: 'if (event == null || event.getMessage() == null) return ctx.getAndClearEmittedEvents()'

File: agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/AguiAgentAdapter.java (hunk 3 above)
Comment (AgentScopeJavaBot) [reply agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/AguiAgentAdapter.java]: ✅ Verified as addressed in `6f36b22`: converters field is now declared 'final' and initialized with Collections.unmodifiableMap(tempConverters) in the constructor

File: agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/StreamContext.java (hunk 5 above)
Comment (AgentScopeJavaBot) [reply agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/StreamContext.java:30]: ✅ Verified as addressed in `6f36b22`: StreamContextTest.java exists with comprehensive coverage: emit/flushEndEvent normal path, flushAllRemainingDeferred with/without deferred events, activeIds add/remove/isActive for text/reasoning/tool

File: agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/AguiAdapterConfig.java (hunk 8 above)
Comment (AgentScopeJavaBot) [reply agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/AguiAdapterConfig.java:261]: ✅ Verified as addressed in `6f36b22`: registerConverter Javadoc now explicitly documents override behavior: 'This allows users to override the default conversion strategies for specific ContentBlock types'

File: agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/AguiAgentAdapter.java (hunk 9 above)
Comment (AgentScopeJavaBot) [reply agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/AguiAgentAdapter.java:127]: ✅ Verified as addressed in `6f36b22`: Praise comment; code still matches description — StreamContext is created inside Flux.defer ensuring fresh context per subscription

File: agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/StreamContext.java (hunk 10 above)
Comment (AgentScopeJavaBot) [reply agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/StreamContext.java:115]: ✅ Verified as addressed in `6f36b22`: Praise comment; code still matches description — LinkedHashMap for deferredEndEvents with prefixed keys (text:/reasoning:/tool:) and flushEndEvent/flushAllRemainingDeferred pattern intact

File: agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/strategy/BlockEventConverter.java (hunk 11 above)
Comment (AgentScopeJavaBot) [reply agentscope-extensions/agentscope-extensions-agui/src/main/java/io/agentscope/core/agui/adapter/strategy/BlockEventConverter.java:31]: ✅ Verified as addressed in `6f36b22`: Praise comment; BlockEventConverter interface still has supportedBlockType() + isApplicable() cleanly separating type matching from applicability checking

**Review by AgentScopeJavaBot — APPROVED** [review]
All blocking review comments have been addressed. 3 non-blocking suggestion(s) (minor/nit) remain open and may be addressed in a follow-up.
